README
This is a source file distribution for the game dungeon as implemented
in C.  It is based on the game dungeon as distributed on a DECUS tape,
circa 1980.  It has been converted from the original DEC FORTRAN to
f77 to C.  See the file "History" for some revision history and credit
to those whose efforts have made this possible.

Take a look at the Makefile.  It should be fine for most systems,
although you may want to change BINDIR and LIBDIR.  On SCO UNIX see
the note at the definition of CFLAGS.  Makefile.MSC should work for
MS-DOS using Microsoft C.  It was contributed by Jonathan Mark
(uunet!microsoft!jonm).

To compile and link dungeon, type make.  To install it in BINDIR and
LIBDIR, type make install.

There are two functions in local.c that you may want to write for your
system.  The first controls when the game can be played, and can be
used to disallow play during business hours, for example.  The second
controls who is allowed to invoke the game debugging tool; note that
this will only be available at all if you uncomment the GDTFLAG line
in the Makefile.  The comments in local.c explain what to do.

All files in the distribution kit are ASCII.  The files dtextc.uu1,
dtextc.uu2, dtextc.uu3, dtextc.uu4 are parts of a uuencoded binary
file named dtextc.dat.  The Makefile will create the binary file
automatically on a UNIX system; elsewhere you will have to stick the
four files together in numerical order and run the resulting large
file through uudecode.  I can't help you find uudecode, though.

The binary file dtextc.dat holds the text strings and initialization
information for the game.  The strings are encrypted to prevent easy
cheating; if you want to do further work on the program, or translate
the strings, Ian Taylor (address below) has a program to convert this
file back and forth from a human-readable form.

This has been compiled and tested on a DECstation 3100 running Ultrix
4.0, a VAXstation GPX running Ultrix 3.1, an 80386 box running SCO
Unix 3.2.2, an 8800 running Ultrix, a Sun box running SUN OS 4 release
4, and an 80386 PC running MS-DOS.

I consider my changes to be in the public domain, as did previous
contributors (see the History file for more detail).  The original
source, however, is copyright.

	Ian Lance Taylor
	ian@airs.com or uunet!airs!ian
	11 March 1991

Dungeon consists of the following files:

	dmain.c			-program root
	dgame.c			-main routine
	dsub.c			-resident subroutines
	dinit.c			-initialization routine
	np.c			-parser, part 0
	np1.c			-parser, part 1
	np2.c			-parser, part 2
	np3.c			-parser, part 3
	gdt.c			-game debugging tool
	verbs.c			-principal verbs
	objcts.c		-principal objects
	sverbs.c		-simple verbs
	dverb1.c		-auxiliary verbs, part 1
	dverb2.c		-auxiliary verbs, part 2
	actors.c		-character processors
	demons.c		-demon processors
	clockr.c		-clock event processors
	rooms.c			-room processors
	nrooms.c		-new room processors
	sobjs.c			-simple objects
	nobjs.c			-new objects
	ballop.c		-balloon processor
	lightp.c		-light processors
	villns.c		-villain processors
	dso1.c			-overlaid subroutines, part 1
	dso2.c			-overlaid subroutines, part 2
	dso3.c			-overlaid subroutines, part 3
	dso4.c			-overlaid subroutines, part 4
	dso5.c			-overlaid subroutines, part 5
	dso6.c			-overlaid subroutines, part 6
	dso7.c			-overlaid subroutines, part 7
	funcs.h			-header file with function prototypes
	vars.h			-header file with variable definitions
	parse.h			-header file for parsing routines
	supp.c			-support routines and more processing
	local.c			-hooks for local definition
	dindxc.dat		-data base [binary file]
	dungeon.6		-man page
	Makefile		-Makefile
	Makefile.MSC		-Makefile for MS-DOS and Microsoft C
	History			-Some revision history notes
	README			-This file

------------------------------------------------------------------------------

I supplied a new file Makefile.BCC for Borland C++ (tested with version 3.1).
While working on this, I discovered that I had to change function main in
dmain.c from 'returning void' to 'returning int' - bcc doesn't accept
'void main(...)' in ANSI mode.

Volker Blasius, 11jul93                                <Volker.Blasius@gmd.de>


history
	History of the C Implementation of Dungeon
	==========================================

This version of dungeon has been modified from FORTRAN to C.  The
original was written in DEC FORTRAN, translated from MDL.  It was then
translated to f77 for UN*X systems, from which it was translated to C.
The C translation was done with the help of f2c, the FORTRAN to C
translator written by David Gay (AT&T Bell Labs), Stu Feldman
(Bellcore), Mark Maimone (Carnegie-Mellon University), and Norm
Schryer (AT&T Bell Labs).

I. From the original documentation...

To:	Dungeon Players
From:	"The Translator"
Subj:	Game Information
Date:	8-OCT-80


This is the first (and last) source release of the PDP-11 version of 
Dungeon.

Please note that Dungeon has been superceded by the game ZORK(tm).
The following is an extract from the new product announcement for
ZORK in the September, 1980 issue of the RT-11 SIG newsletter:

  "'ZORK:  The Great Underground Empire - Part I' ...was developed
   by the original authors based on their ZORK (Dungeon) game for
   the PDP-10.  It features a greatly improved parser;  command
   input and transcript output files;  SAVEs to any device and
   file name;  and adaptation to different terminal types,
   including a status line on VT100s.  Note:  this is not the
   FORTRAN version that has been available through DECUS.  This
   version has been completely rewritten to run efficiently on
   small machines - up to 10 times as fast as the DECUS version.

   ...ZORK runs under RT-ll, HT-ll, or RSTS/E and requires as
   little as 20K words of memory and a single floppy disk drive.
   The game package, consisting of an RX01-format diskette and
   an instruction booklet, is available from Infocom, Inc.,
   P.O. Box 120, Kendall Station, Cambridge, Ma. 02142."

ZORK(tm) is a trademark of Infocom, Inc.  It is available for several
popular personal computers as well as for the PDP-ll.


SUMMARY
-------

		    Welcome to Dungeon!

   Dungeon is a game of adventure, danger, and low cunning.  In it
you will explore some of the most amazing territory ever seen by mortal
man.  Hardened adventurers have run screaming from the terrors contained
within.

   In Dungeon, the intrepid explorer delves into the forgotten secrets
of a lost labyrinth deep in the bowels of the earth, searching for
vast treasures long hidden from prying eyes, treasures guarded by
fearsome monsters and diabolical traps!

   No DECsystem should be without one!

   Dungeon was created at the Programming Technology Division of the MIT
Laboratory for Computer Science by Tim Anderson, Marc Blank, Bruce
Daniels, and Dave Lebling.  It was inspired by the Adventure game of
Crowther and Woods, and the Dungeons and Dragons game of Gygax
and Arneson.  The original version was written in MDL (alias MUDDLE).
The current version was translated from MDL into FORTRAN IV by
a somewhat paranoid DEC engineer who prefers to remain anonymous.

   On-line information may be obtained with the commands HELP and INFO.

II. DEC FORTRAN to f77 Conversion (17-nov-81)

The conversion from DEC FORTRAN to Unix f77 was done by Randy
Dietrich, Lynn Cochran and Sig Peterson.  Much hacking was done to get
it to fit in the limited address space of a PDP-11/44 (split I/D).
Suffice it to say that by leaving out the debugging package and not
linking in the f77 i/o library they managed to get it to run.

III. PDP to VAX (dec-85)

Based on the work of Randy, Lynn and Sig, Bill Randle folded in the
full save/restore functions and the game debugging package (gdt) into
the pdp version to create a Vax/Unix version.  This version also uses
f77 i/o, thus eliminating the extra speak and listen processes needed
on the pdp.

IV. Cleanup I (11-dec-86)

John Gilmore (hoptoad!gnu) cleaned up the source files by moving
most of the common declarations into include files and added
comments from the original (FORTRAN or MDL?) source.  His efforts
are greatly appreciated.

V. Cleanup II (9-feb-87)

Bill Randle (billr@tekred.tek.com) added the pdp dependencies back
into the Vax source files with #ifdefs in order to have just one
set of sources.  Previously, there were two sets of source: one for
the pdp and one for the Vax.  In addition, a shell escape of the
form !cmd was added and the wizard can enter the gdt without having
to recompile the source.  Finally, a man page was generated, based
on the dungeon.doc file.

VI. f77 to C (11-mar-91)

Ian Lance Taylor (ian@airs.com or uunet!airs!ian) used the f2c
translator to generate C source code.  The resulting code was modified
to remove the FORTRAN I/O library, to add simple more processing, and
to change the format of the database file.  Andre Srinivasan
(andre@cs.pitt.edu) help test it.  Jonathan Mark
(uunet!microsoft!jonm) made it work under MS-DOS and Microsoft C.


gdt.c
/* GDT- GAME DEBUGGING TOOL */

/*COPYRIGHT 1980, INFOCOM COMPUTERS AND COMMUNICATIONS, CAMBRIDGE MA. 02142*/
/* ALL RIGHTS RESERVED, COMMERCIAL USAGE STRICTLY PROHIBITED */
/* WRITTEN BY R. M. SUPNIK */

#include <stdio.h>
#include <ctype.h>
#include "funcs.h"
#include "vars.h"
#include "parse.h"

#ifdef ALLOW_GDT

void gdt_()
{
    /* Initialized data */

    const integer cmdmax = 38;
    const char *dbgcmd =
	    "DRDODADCDXDHDLDVDFDSAFHENRNTNCNDRRRTRCRDTKEXARAOAAACAXAVD2DNANDMDTAHDPPDDZAZ";
    static const integer argtyp[38] = { 2,2,2,2,2,0,0,2,2,0,1,0,0,0,0,0,0,
	    0,0,0,1,0,3,3,3,3,1,3,2,2,1,2,1,0,0,0,0,1 };

    /* System generated locals */
    integer i__1, i__2;

    /* Local variables */
    integer i, j, k, l, l1;
    char cmd[3];
    integer fmax, smax;
    char buf[80];
    char *z;

/* FIRST, VALIDATE THAT THE CALLER IS AN IMPLEMENTER. */

    fmax = 46;
/* 						!SET ARRAY LIMITS. */
    smax = 22;

    if (debug_1.gdtflg != 0) {
	goto L2000;
    }
/* 						!IF OK, SKIP. */
    more_output("You are not an authorized user.");
/* 						!NOT AN IMPLEMENTER. */
    return;
/* 						!BOOT HIM OFF */

/* GDT, PAGE 2A */

/* HERE TO GET NEXT COMMAND */

L2000:
    printf("GDT>");
/* 						!OUTPUT PROMPT. */
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    cmd[0] = ' ';
    cmd[1] = ' ';
    sscanf(buf, "%2s", cmd);
/* 						!GET COMMAND. */
    if (cmd[0] == '\0')
	goto L2000;
/* 						!IGNORE BLANKS. */
    if (islower(cmd[0]))
	cmd[0] = toupper(cmd[0]);
    if (islower(cmd[1]))
	cmd[1] = toupper(cmd[1]);
    i__1 = cmdmax;
    for (i = 1; i <= i__1; ++i) {
/* 						!LOOK IT UP. */
	if (cmd[0] == dbgcmd[(i - 1) << 1] &&
	    cmd[1] == dbgcmd[((i - 1) << 1) + 1]) {
	    goto L2300;
	}
/* 						!FOUND? */
/* L2100: */
    }
L2200:
    more_output("?");
/* 						!NO, LOSE. */
    goto L2000;

/* L230: */
/* L240: */
/* L225: */
/* L235: */
/* L245: */

L2300:
    switch (argtyp[i - 1] + 1) {
	case 1:  goto L2400;
	case 2:  goto L2500;
	case 3:  goto L2600;
	case 4:  goto L2700;
    }
/* 						!BRANCH ON ARG TYPE. */
    goto L2200;
/* 						!ILLEGAL TYPE. */

L2700:
    printf("Idx,Ary:  ");
/* 						!TYPE 3, REQUEST ARRAY COORDS. */
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    for (z = buf; *z != '\0'; z++)
	if (*z == ',')
	    *z = ' ';
    j = 0;
    k = 0;
    sscanf(buf, "%d %d", &j, &k);
    goto L2400;

L2600:
    printf("Limits:   ");
/* 						!TYPE 2, READ BOUNDS. */
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    for (z = buf; *z != '\0'; z++)
	if (*z == ',')
	    *z = ' ';
    j = 0;
    k = 0;
    sscanf(buf, "%d %d", &j, &k);
    if (k == 0) {
	k = j;
    }
    goto L2400;

L2500:
    printf("Entry:    ");
/* 						!TYPE 1, READ ENTRY NO. */
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    j = 0;
    sscanf(buf, "%d", &j);

L2400:
    switch (i) {
	case 1:  goto L10000;
	case 2:  goto L11000;
	case 3:  goto L12000;
	case 4:  goto L13000;
	case 5:  goto L14000;
	case 6:  goto L15000;
	case 7:  goto L16000;
	case 8:  goto L17000;
	case 9:  goto L18000;
	case 10:  goto L19000;
	case 11:  goto L20000;
	case 12:  goto L21000;
	case 13:  goto L22000;
	case 14:  goto L23000;
	case 15:  goto L24000;
	case 16:  goto L25000;
	case 17:  goto L26000;
	case 18:  goto L27000;
	case 19:  goto L28000;
	case 20:  goto L29000;
	case 21:  goto L30000;
	case 22:  goto L31000;
	case 23:  goto L32000;
	case 24:  goto L33000;
	case 25:  goto L34000;
	case 26:  goto L35000;
	case 27:  goto L36000;
	case 28:  goto L37000;
	case 29:  goto L38000;
	case 30:  goto L39000;
	case 31:  goto L40000;
	case 32:  goto L41000;
	case 33:  goto L42000;
	case 34:  goto L43000;
	case 35:  goto L44000;
	case 36:  goto L45000;
	case 37:  goto L46000;
	case 38:  goto L47000;
    }
    goto L2200;
/* 						!WHAT??? */
/* GDT, PAGE 3 */

/* DR-- DISPLAY ROOMS */

L10000:
    if (! (j > 0 && j <= rooms_1.rlnt && (k > 0 && k <= rooms_1.rlnt) && j <= 
	    k)) {
	goto L2200;
    }
/* 						!ARGS VALID? */
    more_output("RM#  DESC1  DESC2  EXITS ACTION  VALUE  FLAGS");
/* 						!COL HDRS. */
    i__1 = k;
    for (i = j; i <= i__1; ++i) {
	more_output(NULL);
	printf("%3d", i);
	for (l = 1; l <= 6; ++l)
	    printf(" %6d", eqr[i + l * 200 - 201]);
	printf("\n");

/* L10100: */
    }
    goto L2000;


/* DO-- DISPLAY OBJECTS */

L11000:
    if (! (j > 0 && j <= objcts_1.olnt && (k > 0 && k <= objcts_1.olnt) && j 
	    <= k)) {
	goto L2200;
    }
/* 						!ARGS VALID? */
    more_output("OB# DESC1 DESC2 DESCO ACT FLAGS1 FLAGS2 FVL TVL	  SIZE CAPAC ROOM ADV CON  READ");
/* 						!COL HDRS */
    i__1 = k;
    for (i = j; i <= i__1; ++i) {
	more_output(NULL);
	printf("%3d%6d%6d%6d%4d%7d%7d%4d%4d%6d%6d %4d%4d%4d%6d\n",
		i, eqo[i + 1 * 220 - 221], eqo[i + 2 * 220 - 221],
		eqo[i + 3 * 220 - 221], eqo[i + 4 * 220 - 221],
		eqo[i + 5 * 220 - 221], eqo[i + 6 * 220 - 221],
		eqo[i + 7 * 220 - 221], eqo[i + 8 * 220 - 221],
		eqo[i + 9 * 220 - 221], eqo[i + 10 * 220 - 221],
		eqo[i + 11 * 220 - 221], eqo[i + 12 * 220 - 221],
		eqo[i + 13 * 220 - 221], eqo[i + 14 * 220 - 221]);

/* L11100: */
    }
    goto L2000;


/* DA-- DISPLAY ADVENTURERS */

L12000:
    if (! (j > 0 && j <= advs_1.alnt && (k > 0 && k <= advs_1.alnt) && j <= k)
	    ) {
	goto L2200;
    }
/* 						!ARGS VALID? */
    more_output("AD#   ROOM  SCORE  VEHIC OBJECT ACTION  STREN  FLAGS");
    i__1 = k;
    for (i = j; i <= i__1; ++i) {
	more_output(NULL);
	printf("%3d", i);
	for (l = 1; l <= 7; ++l)
	    printf(" %6d", eqa[i + (l << 2) -  5]);
	printf("\n");
/* L12100: */
    }
    goto L2000;


/* DC-- DISPLAY CLOCK EVENTS */

L13000:
    if (! (j > 0 && j <= cevent_1.clnt && (k > 0 && k <= cevent_1.clnt) && j 
	    <= k)) {
	goto L2200;
    }
/* 						!ARGS VALID? */
    more_output("CL#   TICK ACTION  FLAG");
    i__1 = k;
    for (i = j; i <= i__1; ++i) {
	more_output(NULL);
	printf("%3d %6d %6d     %c\n", i, eqc[i + 1 * 25 - 26],
		eqc[i + 2 * 25 - 26],
		cevent_1.cflag[i - 1] ? 'T' : 'F');
/* L13100: */
    }
    goto L2000;


/* DX-- DISPLAY EXITS */

L14000:
    if (! (j > 0 && j <= exits_1.xlnt && (k > 0 && k <= exits_1.xlnt) && j <= 
	    k)) {
	goto L2200;
    }
/* 						!ARGS VALID? */
    more_output("  RANGE   CONTENTS");
/* 						!COL HDRS. */
    i__1 = k;
    for (i = j; i <= i__1; i += 10) {
/* 						!TEN PER LINE. */
/* Computing MIN */
	i__2 = i + 9;
	l = min(i__2,k);
/* 						!COMPUTE END OF LINE. */
	more_output(NULL);
	printf("%3d-%3d  ", i, l);
	for (l1 = i; l1 <= l; ++l1)
	    printf("%7d", exits_1.travel[l1 - 1]);
	printf("\n");
/* L14100: */
    }
    goto L2000;


/* DH-- DISPLAY HACKS */

L15000:
    more_output(NULL);
    printf("THFPOS= %d, THFFLG= %c, THFACT= %c\n",
	   hack_1.thfpos, hack_1.thfflg ? 'T' : 'F',
	   hack_1.thfact ? 'T' : 'F');
    more_output(NULL);
    printf("SWDACT= %c, SWDSTA= %d\n", hack_1.swdact ? 'T' : 'F',
	   hack_1.swdsta);
    goto L2000;


/* DL-- DISPLAY LENGTHS */

L16000:
    more_output(NULL);
    printf("R=%d, X=%d, O=%d, C=%d\n", rooms_1.rlnt, exits_1.xlnt,
	   objcts_1.olnt, cevent_1.clnt);
    more_output(NULL);
    printf("V=%d, A=%d, M=%d, R2=%d\n", vill_1.vlnt, advs_1.alnt,
	   rmsg_1.mlnt, oroom2_1.r2lnt);
    more_output(NULL);
    printf("MBASE=%d, STRBIT=%d\n", star_1.mbase, star_1.strbit);
    goto L2000;


/* DV-- DISPLAY VILLAINS */

L17000:
    if (! (j > 0 && j <= vill_1.vlnt && (k > 0 && k <= vill_1.vlnt) && j <= k)
	    ) {
	goto L2200;
    }
/* 						!ARGS VALID? */
    more_output("VL# OBJECT   PROB   OPPS   BEST  MELEE");
/* 						!COL HDRS */
    i__1 = k;
    for (i = j; i <= i__1; ++i) {
	more_output(NULL);
	printf("%3d", i);
	for (l = 1; l <= 5; ++l)
	    printf(" %6d", eqv[i + (l << 2) - 5]);
	printf("\n");
/* L17100: */
    }
    goto L2000;


/* DF-- DISPLAY FLAGS */

L18000:
    if (! (j > 0 && j <= fmax && (k > 0 && k <= fmax) && j <= k)) {
	goto L2200;
    }
/* 						!ARGS VALID? */
    i__1 = k;
    for (i = j; i <= i__1; ++i) {
	more_output(NULL);
	printf("Flag #%-2d = %c\n", i, flags[i - 1] ? 'T' : 'F');
/* L18100: */
    }
    goto L2000;


/* DS-- DISPLAY STATE */

L19000:
    more_output(NULL);
    printf("Parse vector= %6d %6d %6d      %c %6d\n",
	   prsvec_1.prsa, prsvec_1.prso, prsvec_1.prsi,
	   prsvec_1.prswon ? 'T' : 'F', prsvec_1.prscon);
    more_output(NULL);
    printf("Play vector=  %6d %6d      %c\n", play_1.winner, play_1.here,
	    play_1.telflg ? 'T' : 'F');
    more_output(NULL);
    printf("State vector= %6d %6d %6d %6d %6d %6d %6d %6d %6d\n",
	   state_1.moves, state_1.deaths, state_1.rwscor, state_1.mxscor,
	   state_1.mxload, state_1.ltshft, state_1.bloc, state_1.mungrm,
	   state_1.hs);
    more_output(NULL);
    printf("              %6d %6d\n", state_1.egscor, state_1.egmxsc);
    more_output(NULL);
    printf("Scol vector=  %6d %6d %6d\n", screen_1.fromdr,
	   screen_1.scolrm, screen_1.scolac);
    goto L2000;

/* GDT, PAGE 4 */

/* AF-- ALTER FLAGS */

L20000:
    if (! (j > 0 && j <= fmax)) {
	goto L2200;
    }
/* 						!ENTRY NO VALID? */
    printf("Old= %c      New= ", flags[j - 1] ? 'T' : 'F');
/* 						!TYPE OLD, GET NEW. */
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    for (z = buf; *z != '\0'; z++) {
	if (! isspace(*z)) {
	    if (*z == 't' || *z == 'T')
		flags[j - 1] = 1;
	    else if (*z == 'f' || *z == 'F')
		flags[j - 1] = 0;
	    break;
	}
    }
    goto L2000;


/* 21000-- HELP */

L21000:
    more_output("Valid commands are:");
    more_output("AA- Alter ADVS          DR- Display ROOMS");
    more_output("AC- Alter CEVENT        DS- Display state");
    more_output("AF- Alter FINDEX        DT- Display text");
    more_output("AH- Alter HERE          DV- Display VILLS");
    more_output("AN- Alter switches      DX- Display EXITS");
    more_output("AO- Alter OBJCTS        DZ- Display PUZZLE");
    more_output("AR- Alter ROOMS         D2- Display ROOM2");
    more_output("AV- Alter VILLS         EX- Exit");
    more_output("AX- Alter EXITS         HE- Type this message");
    more_output("AZ- Alter PUZZLE        NC- No cyclops");
    more_output("DA- Display ADVS        ND- No deaths");
    more_output("DC- Display CEVENT      NR- No robber");
    more_output("DF- Display FINDEX      NT- No troll");
    more_output("DH- Display HACKS       PD- Program detail");
    more_output("DL- Display lengths     RC- Restore cyclops");
    more_output("DM- Display RTEXT       RD- Restore deaths");
    more_output("DN- Display switches    RR- Restore robber");
    more_output("DO- Display OBJCTS      RT- Restore troll");
    more_output("DP- Display parser      TK- Take");
    goto L2000;

/* NR-- NO ROBBER */

L22000:
    hack_1.thfflg = FALSE_;
/* 						!DISABLE ROBBER. */
    hack_1.thfact = FALSE_;
    newsta_(oindex_1.thief, 0, 0, 0, 0);
/* 						!VANISH THIEF. */
    more_output("No robber.");
    goto L2000;

/* NT-- NO TROLL */

L23000:
    findex_1.trollf = TRUE_;
    newsta_(oindex_1.troll, 0, 0, 0, 0);
    more_output("No troll.");
    goto L2000;

/* NC-- NO CYCLOPS */

L24000:
    findex_1.cyclof = TRUE_;
    newsta_(oindex_1.cyclo, 0, 0, 0, 0);
    more_output("No cyclops.");
    goto L2000;

/* ND-- IMMORTALITY MODE */

L25000:
    debug_1.dbgflg = 1;
    more_output("No deaths.");
    goto L2000;

/* RR-- RESTORE ROBBER */

L26000:
    hack_1.thfact = TRUE_;
    more_output("Restored robber.");
    goto L2000;

/* RT-- RESTORE TROLL */

L27000:
    findex_1.trollf = FALSE_;
    newsta_(oindex_1.troll, 0, rindex_1.mtrol, 0, 0);
    more_output("Restored troll.");
    goto L2000;

/* RC-- RESTORE CYCLOPS */

L28000:
    findex_1.cyclof = FALSE_;
    findex_1.magicf = FALSE_;
    newsta_(oindex_1.cyclo, 0, rindex_1.mcycl, 0, 0);
    more_output("Restored cyclops.");
    goto L2000;


/* RD-- MORTAL MODE */

L29000:
    debug_1.dbgflg = 0;
    more_output("Restored deaths.");
    goto L2000;

/* GDT, PAGE 5 */

/* TK-- TAKE */

L30000:
    if (! (j > 0 && j <= objcts_1.olnt)) {
	goto L2200;
    }
/* 						!VALID OBJECT? */
    newsta_(j, 0, 0, 0, play_1.winner);
/* 						!YES, TAKE OBJECT. */
    more_output("Taken.");
/* 						!TELL. */
    goto L2000;


/* EX-- GOODBYE */

L31000:
    prsvec_1.prscon = 1;
    return;

/* AR--	ALTER ROOM ENTRY */

L32000:
    if (! (j > 0 && j <= rooms_1.rlnt && (k > 0 && k <= 5))) {
	goto L2200;
    }
/* 						!INDICES VALID? */
    printf("Old = %6d      New = ", eqr[j + k * 200 - 201]);
/* 						!TYPE OLD, GET NEW. */
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    sscanf(buf, "%d", &eqr[j + k * 200 - 201]);
    goto L2000;


/* AO-- ALTER OBJECT ENTRY */

L33000:
    if (! (j > 0 && j <= objcts_1.olnt && (k > 0 && k <= 14))) {
	goto L2200;
    }
/* 						!INDICES VALID? */
    printf("Old = %6d      New = ", eqo[j + k * 200 - 201]);
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    sscanf(buf, "%d", &eqo[j + k * 220 - 221]);
    goto L2000;

/* AA-- ALTER ADVS ENTRY */

L34000:
    if (! (j > 0 && j <= advs_1.alnt && (k > 0 && k <= 7))) {
	goto L2200;
    }
/* 						!INDICES VALID? */
    printf("Old = %6d      New = ", eqa[j + (k << 2) - 5]);
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    sscanf(buf, "%d", &eqa[j + (k << 2) - 5]);
    goto L2000;

/* AC-- ALTER CLOCK EVENTS */

L35000:
    if (! (j > 0 && j <= cevent_1.clnt && (k > 0 && k <= 3))) {
	goto L2200;
    }
/* 						!INDICES VALID? */
    if (k == 3) {
	goto L35500;
    }
/* 						!FLAGS ENTRY? */
    printf("Old = %6d      New = ", eqc[j + k * 25 - 26]);
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    sscanf(buf, "%d", &eqc[j + k * 25 - 26]);
    goto L2000;

L35500:
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    for (z = buf; *z != '\0'; z++) {
	if (! isspace(*z)) {
	    if (*z == 't' || *z == 'T')
		cevent_1.cflag[j - 1] = 1;
	    else if (*z == 'f' || *z == 'F')
		cevent_1.cflag[j - 1] = 0;
	    break;
	}
    }
    goto L2000;
/* GDT, PAGE 6 */

/* AX-- ALTER EXITS */

L36000:
    if (! (j > 0 && j <= exits_1.xlnt)) {
	goto L2200;
    }
/* 						!ENTRY NO VALID? */
    printf("Old= %6d     New= ", exits_1.travel[j - 1]);
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    sscanf(buf, "%d", &exits_1.travel[j - 1]);
    goto L2000;


/* AV-- ALTER VILLAINS */

L37000:
    if (! (j > 0 && j <= vill_1.vlnt && (k > 0 && k <= 5))) {
	goto L2200;
    }
/* 						!INDICES VALID? */
    printf("Old = %6d      New= ", eqv[j + (k << 2) - 5]);
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    sscanf(buf, "%d", &eqv[j + (k << 2) - 5]);
    goto L2000;

/* D2-- DISPLAY ROOM2 LIST */

L38000:
    if (! (j > 0 && j <= oroom2_1.r2lnt && (k > 0 && k <= oroom2_1.r2lnt) && 
	    j <= k)) {
	goto L2200;
    }
    i__1 = k;
    for (i = j; i <= i__1; ++i) {
	more_output(NULL);
	printf("#%2d   Room=%6d   Obj=%6d\n", i,
		oroom2_1.rroom2[i - 1], oroom2_1.oroom2[i - 1]);
/* L38100: */
    }
    goto L2000;


/* DN-- DISPLAY SWITCHES */

L39000:
    if (! (j > 0 && j <= smax && (k > 0 && k <= smax) && j <= k)) {
	goto L2200;
    }
/* 						!VALID? */
    i__1 = k;
    for (i = j; i <= i__1; ++i) {
	more_output(NULL);
	printf("Switch #%-2d = %d\n", i, switch_[i - 1]);
/* L39100: */
    }
    goto L2000;


/* AN-- ALTER SWITCHES */

L40000:
    if (! (j > 0 && j <= smax)) {
	goto L2200;
    }
/* 						!VALID ENTRY? */
    printf("Old= %6d      New= ", switch_[j - 1]);
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    sscanf(buf, "%d", &switch_[j - 1]);
    goto L2000;

/* DM-- DISPLAY MESSAGES */

L41000:
    if (! (j > 0 && j <= rmsg_1.mlnt && (k > 0 && k <= rmsg_1.mlnt) && j <= k)
	    ) {
	goto L2200;
    }
/* 						!VALID LIMITS? */
    more_output("  RANGE   CONTENTS");
    i__1 = k;
    for (i = j; i <= i__1; i += 10) {
	more_output(NULL);
/* Computing MIN */
	i__2 = i + 9;
	l = min(i__2,k);
	printf("%3d-%3d  ", i, l);
	for (l1 = i; l1 <= l; ++l1)
	    printf(" %6d", rmsg_1.rtext[l1 - 1]);
	printf("\n");
/* L41100: */
    }
    goto L2000;


/* DT-- DISPLAY TEXT */

L42000:
    rspeak_(j);
    goto L2000;

/* AH--	ALTER HERE */

L43000:
    printf("Old= %6d      New= ", play_1.here);
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    sscanf(buf, "%d", &play_1.here);
    eqa[0] = play_1.here;
    goto L2000;

/* DP--	DISPLAY PARSER STATE */

L44000:
    more_output(NULL);
    printf("ORPHS= %7d%7d%7d%7d%7d%7d\n",
	   orp[0], orp[1], orp[2], orp[3], orp[4], last_1.lastit);
    more_output(NULL);
    printf("PV=    %7d%7d%7d%7d%7d\n",
	   pvec[0], pvec[1], pvec[2], pvec[3], pvec[4]);
    more_output(NULL);
    printf("SYN=   %7d%7d%7d%7d%7d%7d\n",
	   syn[0], syn[1], syn[2], syn[3], syn[4], syn[5]);
    more_output(NULL);
    printf("              %7d%7d%7d%7d%7d\n",
	   syn[6], syn[7], syn[8], syn[9], syn[10]);
    goto L2000;


/* PD--	PROGRAM DETAIL DEBUG */

L45000:
    printf("Old= %6d      New= ", debug_1.prsflg);
/* 						!TYPE OLD, GET NEW. */
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    sscanf(buf, "%d", &debug_1.prsflg);
    goto L2000;

/* DZ--	DISPLAY PUZZLE ROOM */

L46000:
    for (i = 1; i <= 64; i += 8) {
/* 						!DISPLAY PUZZLE */
	more_output(NULL);
	printf(" ");
	for (j = i; j <= i + 7; ++j)
	     printf("%3d", puzzle_1.cpvec[j - 1]);
	printf("\n");
/* L46100: */
    }
    goto L2000;


/* AZ--	ALTER PUZZLE ROOM */

L47000:
    if (! (j > 0 && j <= 64)) {
	goto L2200;
    }
/* 						!VALID ENTRY? */
    printf("Old= %6d      New= ", puzzle_1.cpvec[j - 1]);
/* 						!OUTPUT OLD, */
    (void) fflush(stdout);
    (void) fgets(buf, sizeof buf, stdin);
    more_input();
    sscanf(buf, "%d", &puzzle_1.cpvec[j - 1]);
    goto L2000;

} /* gdt_ */

#endif /* ALLOW_GDT */
